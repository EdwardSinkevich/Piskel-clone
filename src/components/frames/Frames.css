.preview-list-wrapper {
  padding-right: 8px;
}

::-webkit-scrollbar {
  background: #CCF6ED;
  border-radius: 4px;
  height: 8px;
  width: 8px;
}

::-webkit-scrollbar-thumb {
  background: #21BB9A;
  border-radius: 4px;
}

.frames-scroller {
  overflow-y: scroll;
  overflow-x: hidden;
  height: 100%;
}

.frame-item {
  position: relative;
  border: #444 3px solid;
  border-radius: 3px;
  margin: 5px 0;
  width: 96px;
  height: 96px;
  z-index: 100;
}

.draggable-mirror {
  display: list-item !important;
}

.frame-item:hover .tile-overlay {
  opacity: 1.0;
}

.tile-view {
  position: relative;
}

.tile-overlay {
  position: absolute;
  background-color: rgba(100, 100, 100, 0.6);
  opacity: 0;
  height: 30px;
  width: 30px;
  border: none;
  cursor: pointer;
}

.icon-frame-duplicate {
  bottom: 0;
  right: 0;
  background-image: url('./images/icons.png');
  background-position: -274px -76px;
}

.icon-frame-recyclebin {
  display: none;
  top: 0;
  right: 0;
  background-image: url('./images/icons.png');
  background-position: -274px -106px;
}

.tile-count {
  top: 0;
  left: 0;
  opacity: 1.0;
  background-color: gold;
  border-bottom-right-radius: 3px;
  font-weight: bold;
  cursor: default;
}

.preview-list {
  padding-right: 9px;
}

.icon-frame-dragndrop {
  display: none;
  bottom: 0;
  left: 0;
  cursor: move;
  background-image: url('./images/icons.png');
  background-position: -274px -46px;
}

.add-frame-action {
  position: relative;
  display: flex;
  align-items: center;
  margin: 10px 0px;
  padding: 6px 0;
  width: 96px;
  border: #888888 solid 3px;
  border-radius: 4px;
  color: #888888;
  background-color: #222222;
  font-size: 13px;
  cursor: pointer;
}

.icon-frame-plus {
  width: 26px;
  height: 26px;
  margin: 3px;
  background-image: url('./images/icons.png');
  background-position: -274px -157px;
  flex-shrink: 0;
}

.label {
  text-align: left;
}

.add-frame-tooltip {
  top: 47px;
  right: 12px;
}
