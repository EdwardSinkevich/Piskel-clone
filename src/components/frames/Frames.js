import './Frames.css';
import { Sortable } from 'sortablejs';

export default class Frames {
  constructor(mainCanvas) {
    this.previewWrapper = null;
    this.previewList = null;
    this.currentFrameItem = null;
    this.amountFrames = 0;
    this.mainCanvasObj = mainCanvas;
    this.mainCanvas = mainCanvas.canvas;
    this.mainCanvasCtx = mainCanvas.canvasCtx;
    this.mainCanvasWrapper = mainCanvas.mainCanvasWrapper;
    this.imgsArray = [];
    this.dragSrcEl = null;
  }

  createFrame(numberOfFrame) {
    const frame = document.createElement('li');
    frame.className = 'frame-item';
    frame.dataset.tileNumber = numberOfFrame;
    frame.innerHTML = `<div class="lowcont-dark-canvas-background">
    <div class="canvas-background"></div>
    <canvas class="tile-view canvas" width="96" height="96"></canvas></div>
    <button class="tile-overlay icon-frame-duplicate" data-tile-action="clone"></button>
    <button class="tile-overlay icon-frame-recyclebin visible" data-tile-action="delete"></button>
    <div class="tile-overlay icon-frame-dragndrop handle visible"></div>
    <button class="tile-overlay tile-count">${numberOfFrame}</button>`;

    this.currentFrameItem = frame;
    this.amountFrames += 1;

    return frame;
  }

  addFrame() {
    const newFrame = this.createFrame(this.amountFrames + 1);
    this.previewList.appendChild(newFrame);

    this.updateAfterAddFrame(this.amountFrames);

    if (this.amountFrames === 1 || this.amountFrames === 2) {
      this.showOrHideBtns();
    }
  }

  createNewImage() {
    const newImg = new Image();

    const drawCurrnetImgToCurrentFrameWrapper = this.drawCurrnetImgToCurrentFrame.bind(this);
    newImg.addEventListener('load', drawCurrnetImgToCurrentFrameWrapper);

    this.mainCanvasObj.currentMainCanvasImg = newImg;
  }

  updateAfterAddFrame(numberOfAddedFrame) {
    this.mainCanvasCtx.clearRect(0, 0, this.mainCanvas.width, this.mainCanvas.height);

    this.createNewImage();

    this.imgsArray.splice(numberOfAddedFrame, 0, this.mainCanvasObj.currentMainCanvasImg);
  }

  removeFrame(removedFrame, removedFrameNumber) {
    this.previewList.removeChild(removedFrame);

    if (removedFrameNumber - 1 === this.imgsArray.indexOf(this.mainCanvasObj.currentMainCanvasImg)) {
      this.updateIfRemoveLastFrame(removedFrameNumber);
    }

    this.imgsArray.splice(removedFrameNumber - 1, 1);

    this.amountFrames -= 1;
    this.updateFramesNumbers(removedFrameNumber - 1);

    if (this.amountFrames === 1) {
      this.showOrHideBtns();
    }
  }

  updateIfRemoveLastFrame(removedFrameNumber) {
    this.currentFrameItem = this.previewList.querySelectorAll('.frame-item')[removedFrameNumber - 2];

    if (this.currentFrameItem) {
      this.mainCanvasObj.currentMainCanvasImg = this.imgsArray[removedFrameNumber - 2];
      this.mainCanvasCtx.clearRect(0, 0, this.mainCanvas.width, this.mainCanvas.height);
      this.mainCanvasCtx.drawImage(this.mainCanvasObj.currentMainCanvasImg, 0, 0, this.mainCanvas.width, this.mainCanvas.height);
    } else {
      this.currentFrameItem = this.previewList.querySelectorAll('.frame-item')[removedFrameNumber - 1];
      this.mainCanvasObj.currentMainCanvasImg = this.imgsArray[removedFrameNumber];
      this.mainCanvasCtx.clearRect(0, 0, this.mainCanvas.width, this.mainCanvas.height);
      this.mainCanvasCtx.drawImage(this.mainCanvasObj.currentMainCanvasImg, 0, 0, this.mainCanvas.width, this.mainCanvas.height);
    }
  }

  copyFrame(copiedFrame, numberOfCopiedFrame) {
    const newFrame = this.createFrame(numberOfCopiedFrame + 1);
    copiedFrame.parentNode.insertBefore(newFrame, copiedFrame.nextSibling);

    this.updateFramesNumbers(numberOfCopiedFrame + 1);

    this.updateAfterCopyFrame(numberOfCopiedFrame);

    if (this.amountFrames === 2) {
      this.showOrHideBtns();
    }
  }

  updateAfterCopyFrame(numberOfCopiedFrame) {
    this.updateAfterAddFrame(numberOfCopiedFrame);

    const copyImg = this.imgsArray[numberOfCopiedFrame - 1].src;

    if (this.imgsArray[numberOfCopiedFrame - 1].src) {
      this.mainCanvasObj.currentMainCanvasImg.src = copyImg;
    }

    this.drawCurrnetImgToCurrentFrame();

    this.mainCanvasCtx.drawImage(this.imgsArray[numberOfCopiedFrame - 1], 0, 0, this.mainCanvas.width, this.mainCanvas.height);
  }

  updateFramesNumbers(numberOfCurrentFrame) {
    const listOfFrames = this.previewList.querySelectorAll('.frame-item');
    const listOfFramesNumbers = this.previewList.querySelectorAll('.tile-count');

    for (let i = numberOfCurrentFrame; i < this.amountFrames; i += 1) {
      listOfFrames[i].dataset.tileNumber = i + 1;
      listOfFramesNumbers[i].textContent = i + 1;
    }
  }

  showOrHideBtns() {
    const removeBtn = this.previewList.querySelector('.icon-frame-recyclebin');
    const moveBtn = this.previewList.querySelector('.icon-frame-dragndrop');

    removeBtn.classList.toggle('visible');
    moveBtn.classList.toggle('visible');
  }

  frameActionsByEvent(e) {
    if (e.target.dataset.tileAction === 'add'
    || e.target.parentNode.dataset.tileAction === 'add') {
      this.addFrame();
    }

    if (e.target.dataset.tileAction === 'clone') {
      const numberOfCopyFrame = +e.target.parentNode.dataset.tileNumber;
      this.copyFrame(e.target.parentNode, numberOfCopyFrame);
    }

    if (e.target.dataset.tileAction === 'delete') {
      const removedFrame = e.target.parentNode;
      const removedFrameNumber = +removedFrame.dataset.tileNumber;

      this.removeFrame(removedFrame, removedFrameNumber);
    }
  }

  drawCurrnetImgToCurrentFrame() {
    const currentFrameCanvas = this.currentFrameItem.querySelector('.tile-view');
    const currentFrameCanvasCtx = currentFrameCanvas.getContext('2d');

    currentFrameCanvasCtx.clearRect(0, 0, currentFrameCanvas.width, currentFrameCanvas.height);
    currentFrameCanvasCtx.drawImage(this.mainCanvasObj.currentMainCanvasImg, 0, 0, currentFrameCanvas.width, currentFrameCanvas.height);
  }

  moveImagesInArray(newIndex, oldIndex) {
    const img = this.imgsArray[oldIndex];
    this.imgsArray.splice(oldIndex, 1);
    this.imgsArray.splice(newIndex, 0, img);
  }

  addDragAndDrop() {
    const updateFramesNumbersWrapper = this.updateFramesNumbers.bind(this);
    const moveImagesInArrayWrapper = this.moveImagesInArray.bind(this);

    new Sortable(this.previewList, {
      handle: '.handle',
      animation: 150,

      onUpdate(e) {
        updateFramesNumbersWrapper(0);
        moveImagesInArrayWrapper(e.newIndex, e.oldIndex);
      },
    });
  }

  addListeners() {
    const frameActionsByEventWrapper = this.frameActionsByEvent.bind(this);
    this.previewWrapper.addEventListener('click', frameActionsByEventWrapper);

    this.addDragAndDrop();
  }

  render() {
    this.previewWrapper = document.getElementById('preview-list-wrapper');
    this.previewWrapper.innerHTML = `
    <div class="frames-scroller">
      <ul id="preview-list" class="preview-list"></ul>
      <div id="add-frame-action" class="add-frame-action tooltip-action"  data-tile-action="add">
        <div class="icon-frame-plus"></div>
        <div class="label">Add new frame</div>
        <span class="tooltip add-frame-tooltip">Press N</span>
      </div>
    </div>`;

    this.previewList = document.getElementById('preview-list');

    this.addFrame();

    this.addListeners();
  }
}
