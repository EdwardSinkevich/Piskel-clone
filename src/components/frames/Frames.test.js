import Frames from './Frames';

describe('Frames.constructor', () => {
  it('Should assign this.mainCanvasObj to mainCanvas', () => {
    const canvasObj = {
      canvas: document.createElement('canvas'),
    };
    const frames = new Frames(canvasObj);
    expect(frames.mainCanvasObj).toEqual(canvasObj);
  });
});

describe('Frames.createFrame', () => {
  it('Should return frame HTMLElement', () => {
    const canvas = document.createElement('canvas');
    const frames = new Frames(canvas);
    const testFrame = document.createElement('li');
    const numberOfFrame = 42;
    testFrame.className = 'frame-item';
    testFrame.dataset.tileNumber = numberOfFrame;
    testFrame.innerHTML = `<div class="lowcont-dark-canvas-background">
    <div class="canvas-background"></div>
    <canvas class="tile-view canvas" width="96" height="96"></canvas></div>
    <button class="tile-overlay icon-frame-duplicate" data-tile-action="clone"></button>
    <button class="tile-overlay icon-frame-recyclebin visible" data-tile-action="delete"></button>
    <div class="tile-overlay icon-frame-dragndrop handle visible"></div>
    <button class="tile-overlay tile-count">${numberOfFrame}</button>`;

    expect(frames.createFrame(42)).toEqual(testFrame);
  });
});

describe('Frames.addFrame', () => {
  it('Should add frame to previewList', () => {
    const previewList = document.createElement('div');
    const canvas = document.createElement('canvas');
    const frames = new Frames(canvas);
    frames.updateAfterAddFrame = () => {};
    frames.previewList = previewList;
    frames.addFrame();
    expect(previewList.hasChildNodes()).toBe(true);
  });
});

describe('Frames.createNewImage', () => {
  it('Should assign this.mainCanvasObj.currentMainCanvasImg to new Image object', () => {
    const canvas = document.createElement('canvas');
    const frames = new Frames(canvas);
    frames.createNewImage();
    expect(frames.mainCanvasObj.currentMainCanvasImg).toBeInstanceOf(Image);
  });
});

describe('Frames.updateAfterAddFrame', () => {
  it('Should call createNewImage', () => {
    const myMock = jest.fn();
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
    };
    const frames = new Frames(canvasObj);
    frames.createNewImage = myMock;
    frames.updateAfterAddFrame();
    expect(myMock.mock.invocationCallOrder.length).toBe(1);
  });
});

describe('Frames.removeFrame', () => {
  it('Should removeChild from this.previewList', () => {
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
    };
    const frames = new Frames(canvasObj);
    const previewList = document.createElement('div');
    frames.updateIfRemoveLastFrame = () => {};
    frames.previewList = previewList;
    frames.addFrame();
    frames.removeFrame(previewList.firstChild, 1);
    expect(previewList.hasChildNodes()).toBe(false);
  });

  it('Should call this.updateIfRemoveLastFrame if remove last frame', () => {
    const myMock = jest.fn();
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
    };
    const frames = new Frames(canvasObj);
    const previewList = document.createElement('div');
    // frames.updateIfRemoveLastFrame = () => {};
    frames.previewList = previewList;
    frames.addFrame();
    frames.updateIfRemoveLastFrame = myMock;
    frames.removeFrame(previewList.firstChild, 1);
    expect(myMock.mock.invocationCallOrder.length).toBe(1);
  });

  it('Should call this.showOrHideBtns if this.amountFrames becames 1', () => {
    const myMock = jest.fn();
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
    };
    const frames = new Frames(canvasObj);
    const previewList = document.createElement('div');
    frames.updateIfRemoveLastFrame = () => {};
    frames.updateFramesNumbers = () => {};
    frames.previewList = previewList;
    frames.addFrame();
    frames.showOrHideBtns = myMock;
    frames.amountFrames = 2;
    frames.removeFrame(previewList.firstChild, 1);
    expect(myMock.mock.invocationCallOrder.length).toBe(1);
  });
});

describe('Frames.updateIfRemoveLastFrame', () => {
  it('Should assign currentMainCanvasImg to imgsArray[removedFrameNumber - 2] if remove last frame', () => {
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
      currentMainCanvasImg: null,
    };
    const frames = new Frames(canvasObj);
    frames.mainCanvasCtx.drawImage = () => {};
    const previewList = document.createElement('div');
    frames.previewList = previewList;
    frames.imgsArray = [41, 42, 43, 44];
    frames.addFrame();
    frames.addFrame();
    frames.addFrame();
    frames.updateIfRemoveLastFrame(2);
    expect(frames.mainCanvasObj.currentMainCanvasImg).toEqual(41);
  });

  it('Should assign currentMainCanvasImg to imgsArray[removedFrameNumber - 2] if remove frame before last', () => {
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
      currentMainCanvasImg: null,
    };
    const frames = new Frames(canvasObj);
    frames.mainCanvasCtx.drawImage = () => {};
    const previewList = document.createElement('div');
    frames.previewList = previewList;
    frames.imgsArray[1] = new Image();
    frames.addFrame();
    frames.updateIfRemoveLastFrame(1);
    expect(frames.mainCanvasObj.currentMainCanvasImg).toEqual(new Image());
  });
});

describe('Frames.copyFrame', () => {
  it('Should copy frame and insert before copied', () => {
    const previewList = document.createElement('div');
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
    };
    const frames = new Frames(canvasObj);
    frames.updateFramesNumbers = () => {};
    frames.updateAfterCopyFrame = () => {};

    frames.previewList = previewList;
    frames.addFrame();
    frames.copyFrame(previewList.firstChild, 0);
    expect(previewList.childElementCount).toBe(2);
  });
});

describe('Frames.updateAfterCopyFrame', () => {
  it('Should call updateAfterAddFrame', () => {
    const myMock = jest.fn();
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
    };
    const frames = new Frames(canvasObj);
    frames.drawCurrnetImgToCurrentFrame = () => {};
    frames.imgsArray = [new Image(), new Image()];
    frames.updateAfterAddFrame = myMock;
    frames.updateAfterCopyFrame(1);
    expect(myMock.mock.invocationCallOrder.length).toBe(1);
  });
});

describe('Frames.updateFramesNumbers', () => {
  it('Should update frames numbers', () => {
    const previewList = document.createElement('div');
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
    };
    const frames = new Frames(canvasObj);
    frames.previewList = previewList;
    frames.addFrame();
    frames.addFrame();
    frames.addFrame();
    frames.addFrame();

    frames.removeFrame(frames.previewList.firstChild, 1);
    frames.updateFramesNumbers();
    const numberLastChild = frames.previewList.lastChild.querySelector('.tile-count').innerHTML;
    expect(numberLastChild).toBe('3');
  });
});

describe('Frames.showOrHideBtns', () => {
  it('Should add className visible to removeBtn', () => {
    const previewList = document.createElement('div');
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
    };
    const frames = new Frames(canvasObj);
    frames.previewList = previewList;
    frames.addFrame();
    frames.showOrHideBtns();
    const removeBtn = frames.previewList.querySelector('.icon-frame-recyclebin');
    expect(removeBtn.classList.contains('visible')).toBe(true);
  });
});

describe('Frames.frameActionsByEvent', () => {
  const canvas = document.createElement('canvas');
  const canvasCtx = canvas.getContext('2d');
  const canvasObj = {
    canvas,
    canvasCtx,
  };
  it('Should call Frame.addFrame', () => {
    const myMock = jest.fn();
    const eventTest = {
      target: {
        dataset: {
          tileAction: 'add',
        },
      },
    };
    const frames = new Frames(canvasObj);
    frames.addFrame = myMock;
    frames.frameActionsByEvent(eventTest);
    expect(myMock.mock.invocationCallOrder.length).toBe(1);
  });

  it('Should call Frame.copyFrame', () => {
    const myMock = jest.fn();
    const eventTest = {
      target: {
        dataset: {
          tileAction: 'clone',
        },
        parentNode: {
          dataset: {
            tileAction: 'clone',
          },
        },
      },
    };
    const frames = new Frames(canvasObj);
    frames.copyFrame = myMock;
    frames.frameActionsByEvent(eventTest);
    expect(myMock.mock.invocationCallOrder.length).toBe(1);
  });

  it('Should call Frame.removeFrame', () => {
    const myMock = jest.fn();
    const eventTest = {
      target: {
        dataset: {
          tileAction: 'delete',
        },
        parentNode: {
          dataset: {
            tileAction: 'delete',
          },
        },
      },
    };
    const frames = new Frames(canvasObj);
    frames.removeFrame = myMock;
    frames.frameActionsByEvent(eventTest);
    expect(myMock.mock.invocationCallOrder.length).toBe(1);
  });
});

describe('Frames.moveImagesInArray', () => {
  it('Should move images in this.imgsArray', () => {
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
    };
    const frames = new Frames(canvasObj);
    frames.imgsArray = [41, 42, 43, 44];
    frames.moveImagesInArray(2, 1);
    expect(frames.imgsArray[2]).toEqual(42);
  });
});

describe('Frames.render', () => {
  it('Should render correctly', () => {
    const canvas = document.createElement('canvas');
    const canvasCtx = canvas.getContext('2d');
    const canvasObj = {
      canvas,
      canvasCtx,
    };
    const frames = new Frames(canvasObj);
    frames.addDragAndDrop = () => {};
    const previewWrapper = document.createElement('div');
    previewWrapper.id = 'preview-list-wrapper';
    document.body.appendChild(previewWrapper);
    frames.render();
    expect(document.getElementById('preview-list-wrapper').innerHTML).toMatchSnapshot();
  });
});
