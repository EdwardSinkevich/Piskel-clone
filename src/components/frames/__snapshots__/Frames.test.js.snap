// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frames.render Should render correctly 1`] = `
"
    <div class=\\"frames-scroller\\">
      <ul id=\\"preview-list\\" class=\\"preview-list\\"><li class=\\"frame-item\\" data-tile-number=\\"1\\"><div class=\\"lowcont-dark-canvas-background\\">
    <div class=\\"canvas-background\\"></div>
    <canvas class=\\"tile-view canvas\\" width=\\"96\\" height=\\"96\\"></canvas></div>
    <button class=\\"tile-overlay icon-frame-duplicate\\" data-tile-action=\\"clone\\"></button>
    <button class=\\"tile-overlay icon-frame-recyclebin\\" data-tile-action=\\"delete\\"></button>
    <div class=\\"tile-overlay icon-frame-dragndrop handle\\"></div>
    <button class=\\"tile-overlay tile-count\\">1</button></li></ul>
      <div id=\\"add-frame-action\\" class=\\"add-frame-action tooltip-action\\" data-tile-action=\\"add\\">
        <div class=\\"icon-frame-plus\\"></div>
        <div class=\\"label\\">Add new frame</div>
        <span class=\\"tooltip add-frame-tooltip\\">Press N</span>
      </div>
    </div>"
`;
